### ISSUE TITLE
Prevent exit application without saving

### ISSUE THREAD

**Issue by idmarjr** [brmodeloweb/brmodelo-app#167]
# Summary

## Problem

Currently the user need to explicitly save the file he/she is working on before exit. Since nowadays users consider all web applications have _autosaving_ as a default behavior, the lack of this functionality is causing multiple users to lose their work when closing browser tab or window without save their changes.

## Reality

Unfortunately implement _autosave_ is not a option yet due to the potential increase in servers resources this feature may demand. Since **BR Modelo Web** is still a small open source project generating no profit, we still cannot afford the cost resultant of this implementation.

## Solution

In order to mitigate the problem we need to implement a confirmation screen, asking the user if he/she want to save changes before exit the screen. This confirmation need to be presented in the follow scenarios: 
- User have changes not saved yet
   - User try to exit the edit screen (Ex: Go back to list screen)
   - User try to close browser tab/window
   - User change the browser URL address and try to load the new address

# Desired output

- [ ] ~~Figma prototype with desired behavior~~
- [x] Frontend implementation

**Comment by oicramps** [comment]
@idmarjr Did you already thought about the modal content?
![image](https://user-images.githubusercontent.com/7519115/141698674-9eb8f120-96a4-4c71-a632-8a376bf8d7dc.png)

**Comment by oicramps** [comment]
@miltonbsn @idmarjr I didn't find anything in the `JointJS` library that tells me if there are any changes in the model. Do you know anything? For now, I'm doing this by mapping the `add`, `change`, and `remove` events and setting an `isDirty` flag so I know if something was changed or not. 

This way is enough, or should we try to identify if the actual model is different from the last saved version instead?

**Comment by miltonbsn** [comment]
There are a couple of ways to see if a paper has changes like `paper._updates.count` for example. However as you have to control an external action (the saving) I would do it the same you are planning. Control a variable on graph / paper events and restart always a save action is made. I think you are in the right path...

About your first point, maybe we could suggest a "save and exit" option

**Comment by oicramps** [comment]
> About your first point, maybe we could suggest a "save and exit" option

@miltonbsn, do you mean a third option `Save and exit`?

**Comment by oicramps** [comment]
@idmarjr @miltonbsn We also have an issue here. When we try to leave the page by changing the angular route, we can show whatever component we want. But this angular transition event isn't triggered when we refresh, change the URL or close the browser tab/window. So for this particular scenario, we have to use the [unload event](https://developer.mozilla.org/en-US/docs/Web/API/Window/unload_event), and in this case, we can only change the message but not the design.

**Using the transition event**
![image](https://user-images.githubusercontent.com/7519115/142064970-e4521227-5093-4052-b3c1-b7cc276ec01b.png)

**Using the unload event**
![image](https://user-images.githubusercontent.com/7519115/142065233-0a2a8cd5-734a-48f9-89b8-55e075f12da8.png)
![image](https://user-images.githubusercontent.com/7519115/142065240-57054432-5a79-4bb6-98e4-6d1ef7c062ce.png)

Do you have any idea about how we can overcome this?

**Comment by idmarjr** [comment]
So, as far as I understand, via Angular route we have zero control to show the dialog message when users are leaving the page by:
- Close tab / Window
- Change URL value
- Refresh the page

It means we're only covering the scenario where user is leaving the page by:
- Click in the back arrow in app header
- Use the "Logout" option

If this is the case I would rather prefer to go for [Unload event](https://developer.mozilla.org/en-US/docs/Web/API/Window/unload_event) approach. That said, the question is:

Do we need to use the default browser `alert()` here?
Cant we `preventDefault()` and trigger a custom function to display our own modal?

---

In a different topic, I would vote against use 3 options in this modal:

> **Using the transition event** ![image](https://user-images.githubusercontent.com/7519115/142064970-e4521227-5093-4052-b3c1-b7cc276ec01b.png)

As much as we think more options is better, default user behavior is to quickly act, not read.

I like what you have in this implementation @oicramps . My only feedback is to change `Continuar aqui` copy and use `Cancelar`. This way we have clear options related to the main action:

**Action:** Exit the document
**Option 1:** Cancel action
**Option 2:** Exit without save

**Comment by oicramps** [comment]
Yeah, With Angular, we have control when changing routes only. Using window events, as far as I know, the `unload` event is the only cancellable one, So for this particular event, we have these behaviors:
- `preventDefault()` will prevent showing the alert, and the page will be refreshed.
- The browser default alert will be used, and we can't override this

For instance:
```javascript
if (isDirty) {
	// Shows the confirmation alert. We can only override the message
	return 'Are you sure you want to reload?'
} else {
	// Allow reload without any alert	
	event.preventDefault()
}
```

**Comment by idmarjr** [comment]
I had a quick chat with @miltonbsn and for now let's go with the `unload` solution since it cover all possible scenarios of user lost work without being notified. 

Further we can dig deeper in the topic and find a solution to improve the visual/usability of this alert.
Regarding the message to display:

> Você tem alterações não salvas. Tem certeza que deseja sair sem salvar?

**Comment by oicramps** [comment]
Do you guys think showing the last time the content was saved is also a good improvement? 

> Salvo por último em 17/11/2021 às 11:23:20

**Comment by idmarjr** [comment]
Definitively.
We actually have some cooking prototypes that have this information on screen

![Screen Shot 2021-11-17 at 11 27 58](https://user-images.githubusercontent.com/301545/142183667-7a27b4c9-dff6-4581-883b-d7bba06363c6.png)

Absolutely no blockers from my side to add this now.
The only thing I will suggest is to do this in another pull request :D

**Comment by oicramps** [comment]
Sure! I'll do that after finishing this one. BTW, I'll open the PR for this issue later tonight. Thank you @idmarjr

**Comment by idmarjr** [comment]
Based in a face to face chat with @oicramps , here's what we found and decided:

1. There's no easy solution to cover all cases ([Check all cases here](https://github.com/brmodeloweb/brmodelo-app/issues/167#issuecomment-970688245))
2. Since we're probably moving away from Angular 1.x in a near future, it makes no sense to look for a solution that work on it's stack

Since we cant runaway to deal with two different implementations, let's embrace it but make sure we have this isolated in the code base. This way when the time to improve or re-implement this come, it will be easier to deal with.

With this in mind, here's what we decided to do:

- Use `isDirty` to track _Tab/page close_, _URL change_ and _page refresh_.
This will Display the default browser alert that cannot be customized. Not the perfect solution but at least it prevent the user to lose work without warning
- For app related interactions (_Return to models list_ and  _logout_) let's use the custom made modal that have a more user friendly look and feel.
